tslearn/testing_utils.py

```diff
@@ -0,0 +1,84 @@
+"""
+The :mod:`tslearn.testing_utils` module includes various utilities that can
+be used for testing.
+"""
+
+import tslearn
+
+import pkgutil
+import inspect
+from operator import itemgetter
+
+from sklearn.base import (BaseEstimator, ClassifierMixin, ClusterMixin,
+                          RegressorMixin, TransformerMixin)
+from sklearn.utils.estimator_checks import check_estimator
+
+
+def get_estimators(type_filter='all'):
+    """Return a list of classes that inherit from `sklearn.BaseEstimator`.
+    This code is based on `sklearn,utils.testing.all_estimators`.
+
+    Parameters
+    ----------
+    type_filter : str
+        A value in ['all', 'classifier', 'transformer', 'cluster'] which
+        defines which type of estimators to retrieve
+
+    Returns
+    -------
+    list
+        Collection of estimators of the type specified in `type_filter`
+    """
+
+    def is_abstract(c):
+        if not(hasattr(c, '__abstractmethods__')):
+            return False
+        if not len(c.__abstractmethods__):
+            return False
+        return True
+
+    if type_filter not in ['all', 'classifier', 'transformer', 'cluster']:
+        # TODO: make this exception more specific
+        raise Exception("type_filter should be element of "
+                        "['all', 'classifier', 'transformer', 'cluster']")
+
+    # Walk through all the packages from our base_path and
+    # add all the classes to a list
+    all_classes = []
+    base_path = tslearn.__path__
+    for _, name, _ in pkgutil.walk_packages(path=base_path,
+                                            prefix='tslearn.'):
+        module = __import__(name, fromlist="dummy")
+        all_classes.extend(inspect.getmembers(module, inspect.isclass))
+
+    # Filter out those that are not a subclass of `sklearn.BaseEstimator`
+    all_classes = [c for c in set(all_classes)
+                   if issubclass(c[1], BaseEstimator)]
+    # get rid of abstract base classes
+    all_classes = [c for c in all_classes if not is_abstract(c[1])]
+
+    # Now filter out the estimators that are not of the specified type
+    filters = {
+        'all': [ClassifierMixin, RegressorMixin,
+                TransformerMixin, ClusterMixin],
+        'classifier': [ClassifierMixin],
+        'transformer': [TransformerMixin],
+        'cluster': [ClusterMixin]
+    }[type_filter]
+    filtered_classes = []
+    for _class in all_classes:
+        if any([issubclass(_class[1], mixin) for mixin in filters]):
+            filtered_classes.append(_class)
+
+    # Remove duplicates and return the list of remaining estimators
+    return sorted(set(filtered_classes), key=itemgetter(0))
+
+
+def check_all_estimators():
+    for estimator in get_estimators('all'):
+        print(estimator[0])
+        check_estimator(estimator[1])
+        print('{} is sklearn compliant.'.format(estimator[0]))
+
+
+check_all_estimators()
```

Comment:
No need for that call if it is in a test suite

Reply:
Yes, the tests are run automatically with a bash script. All the functions that are tests should start with `test_` and should raise an error if the test is not successful (using `assert` or a function from `np.testing` or `sklearn.testing` for instance).

There are many folders with tests in scikit-learn, you can have a look at [sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py](https://github.com/scikit-learn/scikit-learn/blob/7813f7efb5b2012412888b69e73d76f2df2b50b6/sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py) for instance.

Reply:
You can also catch errors and warnings with pytest (if you want to test that an error or a warning is raised):
* https://docs.pytest.org/en/latest/assert.html
* https://docs.pytest.org/en/latest/warnings.html